File: build_combined_6ac.mjs
import fs from 'fs';
import path from 'path';

const ROOT = process.cwd();
const CROSS_DIR = '/Users/jacksonelliott/Downloads/6A-6C connections';

function readGraphFromHtml(htmlPath) {
  const html = fs.readFileSync(htmlPath, 'utf8');
  const m = html.match(/<script id="graph-data" type="application\/json">([\s\S]*?)<\/script>/);
  if (!m) throw new Error(`graph-data script not found in ${htmlPath}`);
  return JSON.parse(m[1]);
}

function parseCsvLine(line) {
  const out = [];
  let cur = '';
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && line[i + 1] === '"') {
        cur += '"';
        i++;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (ch === ',' && !inQuotes) {
      out.push(cur);
      cur = '';
    } else {
      cur += ch;
    }
  }
  out.push(cur);
  return out;
}

function parseCsv(text) {
  return text
    .split(/\r?\n/)
    .filter((line) => line.length > 0)
    .map(parseCsvLine);
}

function normalizeIdsFromCell(cell) {
  if (!cell) return [];
  const matches = String(cell).toUpperCase().match(/\d+\s*[A-Z]+/g) || [];
  return [...new Set(matches.map((s) => s.replace(/\s+/g, '')))];
}

function addEdge(edgeMap, a, b, tags) {
  if (!a || !b) return;
  const [s, t] = [a, b].sort();
  const key = `${s}|${t}`;
  if (!edgeMap.has(key)) {
    edgeMap.set(key, { source: s, target: t, tags: new Set() });
  }
  const e = edgeMap.get(key);
  (tags || []).forEach((tag) => {
    const clean = String(tag || '').trim();
    if (clean) e.tags.add(clean);
  });
}

function decorateNodes(baseNodes, course, moduleOffset) {
  return baseNodes.map((n) => ({
    ...n,
    id: `${course}:${n.id}`,
    base_id: n.id,
    course,
    course_module_num: n.module_num,
    module_num: n.module_num + moduleOffset,
    module_title: `${course} M${n.module_num}: ${n.module_title}`,
  }));
}

function decorateLinks(baseLinks, course) {
  return baseLinks.map((l) => ({
    source: `${course}:${l.source}`,
    target: `${course}:${l.target}`,
    tags: Array.isArray(l.tags) ? l.tags : [],
  }));
}

function computePositions(nodes) {
  const width = 1400;
  const height = 1000;
  const left = 120;
  const right = 120;
  const top = 90;
  const bottom = 90;

  const byModule = new Map();
  for (const n of nodes) {
    if (!byModule.has(n.module_num)) byModule.set(n.module_num, []);
    byModule.get(n.module_num).push(n);
  }

  const modules = Array.from(byModule.keys()).sort((a, b) => a - b);
  const rowCount = Math.max(2, modules.length);
  const rowSpacing = (height - top - bottom) / (rowCount - 1);
  const totalWidth = width - left - right;

  modules.forEach((m, idx) => {
    const arr = byModule.get(m);
    arr.sort((a, b) => a.base_id.localeCompare(b.base_id));
    const step = totalWidth / (arr.length + 1);
    const y = (height - bottom) - idx * rowSpacing;
    arr.forEach((n, i) => {
      n.sx = left + (i + 1) * step;
      n.sy = y;
      n.x = n.sx;
      n.y = n.sy;
    });
  });

  // Ring-like fallback positions for compatibility with existing rings mode.
  const cx = width / 2;
  const cy = height / 2;
  const R = Math.min(width, height) * 0.38;
  modules.forEach((m, mi) => {
    const arr = byModule.get(m);
    const a0 = (2 * Math.PI * mi) / modules.length - Math.PI / 2;
    const mx = cx + R * Math.cos(a0);
    const my = cy + R * Math.sin(a0);
    const rMini = Math.max(18, Math.min(55, 210 / Math.sqrt(arr.length + 1)));
    arr.forEach((n, i) => {
      const a = (2 * Math.PI * i) / Math.max(1, arr.length);
      n.rx = mx + rMini * Math.cos(a);
      n.ry = my + rMini * Math.sin(a);
    });
  });
}

function main() {
  const g6a = readGraphFromHtml(path.join(ROOT, 'map_6A.html'));
  const g6c = readGraphFromHtml(path.join(ROOT, 'map_6C.html'));

  const nodes6a = decorateNodes(g6a.nodes, '6A', 0);
  const nodes6c = decorateNodes(g6c.nodes, '6C', 10);
  const allNodes = [...nodes6a, ...nodes6c];
  const nodeSet = new Set(allNodes.map((n) => n.id));

  const edgeMap = new Map();

  for (const l of decorateLinks(g6a.links, '6A')) addEdge(edgeMap, l.source, l.target, l.tags);
  for (const l of decorateLinks(g6c.links, '6C')) addEdge(edgeMap, l.source, l.target, l.tags);

  const crossFiles = fs.readdirSync(CROSS_DIR).filter((f) => f.toLowerCase().endsWith('.csv'));
  for (const file of crossFiles) {
    const mm = file.match(/Module\s+(\d+)_/i);
    if (!mm) continue;
    const moduleNum = Number(mm[1]);
    const rows = parseCsv(fs.readFileSync(path.join(CROSS_DIR, file), 'utf8'));
    if (rows.length === 0) continue;

    const header = rows[0];
    const labelCol = header.findIndex((c) => String(c).trim().toLowerCase() === 'label');
    const c6Col = header.findIndex((c) => String(c).trim().toLowerCase() === '6c connections');
    if (labelCol < 0 || c6Col < 0) continue;

    for (let i = 1; i < rows.length; i++) {
      const row = rows[i];
      const label = String(row[labelCol] || '').trim().toUpperCase();
      if (!label || !/^[A-Z]+$/.test(label)) continue;
      const src = `6A:${moduleNum}${label}`;
      if (!nodeSet.has(src)) continue;
      const targets = normalizeIdsFromCell(row[c6Col]);
      for (const tgtBase of targets) {
        const tgt = `6C:${tgtBase}`;
        if (!nodeSet.has(tgt)) continue;
        addEdge(edgeMap, src, tgt, ['cross-course', '6A-6C']);
      }
    }
  }

  computePositions(allNodes);

  const links = Array.from(edgeMap.values()).map((e) => ({
    source: e.source,
    target: e.target,
    tags: Array.from(e.tags).sort(),
  }));

  const out = { nodes: allNodes, links };
  fs.writeFileSync(path.join(ROOT, 'graph_6AC.json'), JSON.stringify(out, null, 2) + '\n');

  const crossCount = links.filter((l) => l.source.startsWith('6A:') && l.target.startsWith('6C:')).length;
  console.log(`Wrote graph_6AC.json with ${allNodes.length} nodes, ${links.length} links (${crossCount} cross-course).`);
}

main();

File: build_combined_6ac.py
import csv
import json
import math
import re
from pathlib import Path

ROOT = Path.cwd()
CROSS_DIR = Path('/Users/jacksonelliott/Downloads/6A-6C connections')


def read_graph_from_html(html_path: Path):
    text = html_path.read_text(encoding='utf-8')
    m = re.search(r'<script id="graph-data" type="application/json">([\s\S]*?)</script>', text)
    if not m:
        raise RuntimeError(f'graph-data script not found in {html_path}')
    return json.loads(m.group(1))


def normalize_ids_from_cell(cell: str):
    if not cell:
        return []
    vals = re.findall(r'\d+\s*[A-Z]+', cell.upper())
    out = []
    seen = set()
    for v in vals:
        vv = re.sub(r'\s+', '', v)
        if vv not in seen:
            seen.add(vv)
            out.append(vv)
    return out


def add_edge(edge_map, a, b, tags):
    if not a or not b:
        return
    s, t = sorted([a, b])
    key = f'{s}|{t}'
    if key not in edge_map:
        edge_map[key] = {'source': s, 'target': t, 'tags': set()}
    for tag in tags or []:
        clean = str(tag).strip()
        if clean:
            edge_map[key]['tags'].add(clean)


def decorate_nodes(base_nodes, course, module_offset):
    out = []
    for n in base_nodes:
        nn = dict(n)
        nn['id'] = f"{course}:{n['id']}"
        nn['base_id'] = n['id']
        nn['course'] = course
        nn['course_module_num'] = n['module_num']
        nn['module_num'] = n['module_num'] + module_offset
        nn['module_title'] = f"{course} M{n['module_num']}: {n['module_title']}"
        out.append(nn)
    return out


def decorate_links(base_links, course):
    out = []
    for l in base_links:
        out.append({
            'source': f"{course}:{l['source']}",
            'target': f"{course}:{l['target']}",
            'tags': list(l.get('tags') or []),
        })
    return out


def compute_positions(nodes):
    width, height = 1400, 1000
    left, right, top, bottom = 120, 120, 90, 90

    by_module = {}
    for n in nodes:
        by_module.setdefault(n['module_num'], []).append(n)

    modules = sorted(by_module.keys())
    row_count = max(2, len(modules))
    row_spacing = (height - top - bottom) / (row_count - 1)
    total_width = width - left - right

    for idx, m in enumerate(modules):
        arr = by_module[m]
        arr.sort(key=lambda n: n['base_id'])
        step = total_width / (len(arr) + 1)
        y = (height - bottom) - idx * row_spacing
        for i, n in enumerate(arr):
            n['sx'] = left + (i + 1) * step
            n['sy'] = y
            n['x'] = n['sx']
            n['y'] = n['sy']

    cx, cy = width / 2, height / 2
    big_r = min(width, height) * 0.38
    for mi, m in enumerate(modules):
        arr = by_module[m]
        a0 = (2 * math.pi * mi) / len(modules) - math.pi / 2
        mx = cx + big_r * math.cos(a0)
        my = cy + big_r * math.sin(a0)
        mini_r = max(18.0, min(55.0, 210.0 / math.sqrt(len(arr) + 1)))
        k = max(1, len(arr))
        for i, n in enumerate(arr):
            a = (2 * math.pi * i) / k
            n['rx'] = mx + mini_r * math.cos(a)
            n['ry'] = my + mini_r * math.sin(a)


def main():
    g6a = read_graph_from_html(ROOT / 'map_6A.html')
    g6c = read_graph_from_html(ROOT / 'map_6C.html')

    nodes6a = decorate_nodes(g6a['nodes'], '6A', 0)
    nodes6c = decorate_nodes(g6c['nodes'], '6C', 10)
    nodes = nodes6a + nodes6c
    node_ids = {n['id'] for n in nodes}

    edge_map = {}

    for l in decorate_links(g6a['links'], '6A'):
        add_edge(edge_map, l['source'], l['target'], l['tags'])
    for l in decorate_links(g6c['links'], '6C'):
        add_edge(edge_map, l['source'], l['target'], l['tags'])

    for csv_path in sorted(CROSS_DIR.glob('*.csv')):
        mm = re.search(r'Module\s+(\d+)_', csv_path.name, re.IGNORECASE)
        if not mm:
            continue
        module_num = int(mm.group(1))

        with csv_path.open('r', encoding='utf-8', newline='') as f:
            rows = list(csv.reader(f))

        if not rows:
            continue

        label_col = -1
        c6_col = -1
        for r in rows[:4]:
            lowered = [str(c).strip().lower() for c in r]
            if label_col < 0 and 'label' in lowered:
                label_col = lowered.index('label')
            if c6_col < 0 and '6c connections' in lowered:
                c6_col = lowered.index('6c connections')
        if label_col < 0 or c6_col < 0:
            continue

        for row in rows[1:]:
            if label_col >= len(row):
                continue
            label = str(row[label_col]).strip().upper()
            if not re.fullmatch(r'[A-Z]+', label or ''):
                continue
            src = f'6A:{module_num}{label}'
            if src not in node_ids:
                continue
            cell = row[c6_col] if c6_col < len(row) else ''
            for tgt_base in normalize_ids_from_cell(cell):
                tgt = f'6C:{tgt_base}'
                if tgt in node_ids:
                    add_edge(edge_map, src, tgt, ['cross-course', '6A-6C'])

    compute_positions(nodes)

    links = []
    cross_count = 0
    for e in edge_map.values():
        tags = sorted(e['tags'])
        links.append({'source': e['source'], 'target': e['target'], 'tags': tags})
        if e['source'].startswith('6A:') and e['target'].startswith('6C:'):
            cross_count += 1

    out = {'nodes': nodes, 'links': links}
    (ROOT / 'graph_6AC.json').write_text(json.dumps(out, indent=2) + '\n', encoding='utf-8')
    print(f"Wrote graph_6AC.json with {len(nodes)} nodes, {len(links)} links ({cross_count} cross-course).")


if __name__ == '__main__':
    main()
